deployer/helm_upgrade_decision.py

```diff
@@ -0,0 +1,367 @@
+    and/or all clusters appropriately. These common files include the helm charts we
+    deploy, as well as the GitHub Actions and deployer package we use to deploy them.
+    Args:
+        modified_paths (list[str]): The list of files that have been added or modified
+            in a given GitHub Pull Request.
+    Returns:
+        upgrade_all_clusters (bool): Whether or not all clusters should be upgraded
+            since the support chart has changed
+        upgrade_all_hubs (bool): Whether or not all hubs on all clusters should be
+            upgraded since a core piece of infrastructure has changed
+    """
+    # If any of the following filepaths have changed, we should update all hubs on all clusters
+    common_filepaths = [
+        "deployer/*",
+        "helm-charts/basehub/*",
+        "helm-charts/daskhub/*",
+        ".github/actions/deploy/*",
+        "requirements.txt",
+        ".github/workflows/deploy-hubs.yaml",
+    ]
+    # If this filepath has changes, we should update the support chart on all clusters
+    support_chart_filepath = "helm-charts/support/*"
+
+    # Discover if the support chart has been modified
+    upgrade_all_clusters = bool(fnmatch.filter(modified_paths, support_chart_filepath))
+
+    # Discover if any common config has been modified
+    upgrade_all_hubs = False
+    while not upgrade_all_hubs:
+        upgrade_all_hubs = bool(
+            fnmatch.filter(modified_paths, common_filepath_pattern)
+            for common_filepath_pattern in common_filepaths
+        )
+
+    return upgrade_all_clusters, upgrade_all_hubs
+
+
+def get_categorised_unique_filepaths(added_or_modified_files: list):
+    """For a list of added and modified files, get the following categorised, unique
+    filepaths:
+    - A list of unique filepaths to cluster folders containing added/modified files
+    - A set of all added/modified files matching the pattern "*/cluster.yaml"
+    - A set of all added/modified files matching the pattern "*/*.values.yaml"
+    - A set of all added/modified files matching the pattern "*/support.values.yaml"
+
+    Note: "cluster folders" are those that contain a cluster.yaml file.
+
+    Args:
+        added_or_modified_files (list[str]): A list of files that have been added or
+            modified in a GitHub Pull Request
+
+    Returns:
+        list[path obj]: A list of unique filepaths to cluster folders
+        set[str]: A set of all files matching the pattern "*/cluster.yaml"
+        set[str]: A set of all files matching the pattern "*/*.values.yaml"
+        set[str]: A set of all files matching the pattern "*/*support*.values.yaml"
+    """
+    patterns_to_match = ["*/cluster.yaml", "*/*.values.yaml", "*/*support*.values.yaml"]
+    cluster_filepaths = []
+
+    # Identify cluster paths amongst target paths depending on the files they contain
+    for pattern in patterns_to_match:
+        cluster_filepaths.extend(fnmatch.filter(added_or_modified_files, pattern))
```

Comment:
_Original discussion:_ https://github.com/2i2c-org/infrastructure/pull/1093#discussion_r826182520

`patterns_to_matc`h lists three patterns, but the third is matching a subset of the second pattern, making it redundant it seems.

The same patterns show up again later when its relevant to separate the general values file paths from the support file paths. This seems fine, but it breaks DRY principle a bit and could be refactored to something like `support_values_pattern` and `values_pattern` or `non_support_values_pattern`.

Reply:
@consideRatio For this, I really wish `fnmatch.filter()` would output _two_ lists - one with filenames that match, and the other with ones that don't. Maybe we need to write something more bespoke here? 😕

Reply:
Now I think of it as the code accomplishes the intentions of the function, so at this point I figure let's wait with adjusting low level tech details like this.